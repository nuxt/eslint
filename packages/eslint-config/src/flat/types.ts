import type { StylisticCustomizeOptions } from '@stylistic/eslint-plugin'

export interface ToolingOptions {
  /**
   * Enable RegExp rules
   *
   * @see https://github.com/ota-meshi/eslint-plugin-regexp
   * @default true
   */
  regexp?: boolean
  /**
   * Enable Unicorn rules
   *
   * @see https://github.com/sindresorhus/eslint-plugin-unicorn
   * @default true
   */
  unicorn?: boolean
  /**
   * Enable jsdoc rules
   *
   * @default true
   */
  jsdoc?: boolean
}

export interface NuxtESLintFeaturesOptions {
  /**
   * Setup basic JavaScript, TypeScript and Vue plugins and rules.
   *
   * You might want to disable it when you are using other ESLint config that handles the basic setup.
   *
   * @default true
   */
  standalone?: boolean

  /**
   * Enable rules for Nuxt module authors or library authors
   *
   * @experimental Changes might not follow semver
   * @default false
   */
  tooling?: boolean | ToolingOptions

  /**
   * Enable stylistic ESLint rules for formatting and code style check
   *
   * @see https://eslint.style/guide/config-presets
   * @default false
   */
  stylistic?: boolean | StylisticCustomizeOptions<true>

  /**
   * Options for TypeScript setup
   *
   * @default true
   */
  typescript?: boolean | {
    /**
     * Enable strict rules
     * @see https://typescript-eslint.io/users/configs#strict
     * @default true
     */
    strict?: boolean
  }
}

export interface NuxtESLintConfigOptions {
  features?: NuxtESLintFeaturesOptions

  dirs?: {
    /**
     * Nuxt source directory
     */
    src?: string[]

    /**
     * Root directory for nuxt project
     */
    root?: string[]

    /**
     * Directory for pages
     */
    pages?: string[]

    /**
     * Directory for layouts
     */
    layouts?: string[]

    /**
     * Directory for components
     */
    components?: string[]

    /**
     * Directory for components with prefix
     * Ignore `vue/multi-word-component-names`
     */
    componentsPrefixed?: string[]

    /**
     * Directory for composobles
     */
    composables?: string[]

    /**
     * Directory for plugins
     */
    plugins?: string[]

    /**
     * Directory for modules
     */
    modules?: string[]

    /**
     * Directory for middleware
     */
    middleware?: string[]

    /**
     * Directory for server
     */
    servers?: string[]
  }
}

type NotNill<T> = T extends null | undefined ? never : T

export interface NuxtESLintConfigOptionsResolved {
  features: Required<NotNill<NuxtESLintFeaturesOptions>>
  dirs: Required<NotNill<NuxtESLintConfigOptions['dirs']>>
}

export type Awaitable<T> = T | Promise<T>
