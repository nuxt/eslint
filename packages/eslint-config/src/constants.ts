export const GLOB_EXTS = '{js,ts,jsx,tsx,vue}'
