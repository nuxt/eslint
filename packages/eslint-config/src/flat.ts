export * from './flat/index'
export { default } from './flat/index'
