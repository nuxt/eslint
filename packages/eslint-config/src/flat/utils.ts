import type { NuxtESLintConfigOptions, NuxtESLintConfigOptionsResolved } from '../flat'

export function removeUndefined<T extends object>(obj: T): T {
  return Object.fromEntries(Object.entries(obj).filter(([, value]) => value !== undefined)) as T
}

export function resolveOptions(
  config: NuxtESLintConfigOptions,
): NuxtESLintConfigOptionsResolved {
  if ('__resolved' in config) {
    return config as NuxtESLintConfigOptionsResolved
  }

  const dirs = {
    ...config.dirs,
  } as NuxtESLintConfigOptionsResolved['dirs']

  dirs.root ||= ['.']
  dirs.src ||= dirs.root
  dirs.pages ||= dirs.src.map(src => `${src}/pages`)
  dirs.layouts ||= dirs.src.map(src => `${src}/layouts`)
  dirs.components ||= dirs.src.map(src => `${src}/components`)
  dirs.composables ||= dirs.src.map(src => `${src}/composables`)
  dirs.plugins ||= dirs.src.map(src => `${src}/plugins`)
  dirs.modules ||= dirs.src.map(src => `${src}/modules`)
  dirs.middleware ||= dirs.src.map(src => `${src}/middleware`)
  dirs.servers ||= dirs.src.map(src => `${src}/servers`)

  const resolved: NuxtESLintConfigOptionsResolved = {
    features: {
      standalone: true,
      stylistic: false,
      typescript: true,
      tooling: false,
      ...config.features,
    },
    dirs,
  }

  Object.defineProperty(resolved, '__resolved', { value: true, enumerable: false })

  return resolved
}
