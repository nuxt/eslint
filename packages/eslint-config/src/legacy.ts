import '@rushstack/eslint-patch/modern-module-resolution'
import type { ESLint } from 'eslint'

const config: ESLint.ConfigData = {
  parserOptions: {
    ecmaVersion: 'latest',
    parser: {
      js: 'espree',
      jsx: 'espree',
      ts: '@typescript-eslint/parser',
      tsx: '@typescript-eslint/parser',
    },
    extraFileExtensions: ['.vue'],
    ecmaFeatures: {
      jsx: true,
    },
  },
  env: { node: true },
  plugins: ['@typescript-eslint'],
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:vue/vue3-recommended',
  ],
  overrides: [
    {
      files: ['*.ts', '*.tsx', '*.mts', '*.cts', '*.vue'],
      rules: {
        // The core 'no-unused-vars' rules (in the eslint:recommended ruleset)
        // does not work with type definitions.
        'no-unused-vars': 'off',
        '@typescript-eslint/no-unused-vars': 'warn',
      },
    },
    {
      // Include typescript eslint rules in *.vue files
      // https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/configs/eslint-recommended.ts
      files: ['*.vue'],
      rules: {
        'constructor-super': 'off', // ts(2335) & ts(2377)
        'getter-return': 'off', // ts(2378)
        'no-const-assign': 'off', // ts(2588)
        'no-dupe-args': 'off', // ts(2300)
        'no-dupe-class-members': 'off', // ts(2393) & ts(2300)
        'no-dupe-keys': 'off', // ts(1117)
        'no-func-assign': 'off', // ts(2539)
        'no-import-assign': 'off', // ts(2539) & ts(2540)
        'no-new-symbol': 'off', // ts(7009)
        'no-obj-calls': 'off', // ts(2349)
        'no-redeclare': 'off', // ts(2451)
        'no-setter-return': 'off', // ts(2408)
        'no-this-before-super': 'off', // ts(2376)
        'no-undef': 'off', // ts(2304)
        'no-unreachable': 'off', // ts(7027)
        'no-unsafe-negation': 'off', // ts(2365) & ts(2360) & ts(2358)
        'no-var': 'error', // ts transpiles let/const to var, so no need for vars any more
        'prefer-const': 'error', // ts provides better types with const
        'prefer-rest-params': 'error', // ts provides better types with rest args over arguments
        'prefer-spread': 'error', // ts transpiles spread to apply, so no need for manual apply
        'valid-typeof': 'off', // ts(2367)
      },
    },
    {
      files: [
        // These pages are not used directly by users so they can have one-word names.
        '**/pages/**/*.{js,ts,jsx,tsx,vue}',
        '**/layouts/**/*.{js,ts,jsx,tsx,vue}',
        '**/app.{js,ts,jsx,tsx,vue}',
        '**/error.{js,ts,jsx,tsx,vue}',
        // These files should have multiple words in their names as they are within subdirectories.
        '**/components/*/**/*.{js,ts,jsx,tsx,vue}',
      ],
      rules: { 'vue/multi-word-component-names': 'off' },
    },
    {
      // Pages and layouts are required to have a single root element if transitions are enabled.
      files: ['**/pages/**/*.{js,ts,jsx,tsx,vue}', '**/layouts/**/*.{js,ts,jsx,tsx,vue}'],
      rules: { 'vue/no-multiple-template-root': 'error' },
    },
  ],
}

export default config
