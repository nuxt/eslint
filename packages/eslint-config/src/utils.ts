import process from 'node:process'
import { isPackageExists } from 'local-pkg'
import type { NuxtESLintConfigOptions, NuxtESLintConfigOptionsResolved } from './flat'
import type { Awaitable } from './types'

export const parserPlain = {
  meta: {
    name: 'parser-plain',
  },
  parseForESLint: (code: string) => ({
    ast: {
      body: [],
      comments: [],
      loc: { end: code.length, start: 0 },
      range: [0, code.length],
      tokens: [],
      type: 'Program',
    },
    scopeManager: null,
    services: { isPlain: true },
    visitorKeys: {
      Program: [],
    },
  }),
}

export async function ensurePackages(packages: (string | undefined)[]): Promise<void> {
  if (process.env.CI || !process.stdout.isTTY)
    return

  const nonExistingPackages = packages.filter(i => i && !isPackageExists(i)) as string[]
  if (nonExistingPackages.length === 0)
    return

  const p = await import('@clack/prompts')
  const result = await p.confirm({
    message: `${nonExistingPackages.length === 1 ? 'Package is' : 'Packages are'} required for this config: ${nonExistingPackages.join(', ')}. Do you want to install them?`,
  })
  if (result)
    await import('@antfu/install-pkg').then(i => i.installPackage(nonExistingPackages, { dev: true }))
}

export async function interopDefault<T>(m: Awaitable<T>): Promise<T extends { default: infer U } ? U : T> {
  const resolved = await m
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return (resolved as any).default || resolved
}

export function removeUndefined<T extends object>(obj: T): T {
  return Object.fromEntries(Object.entries(obj).filter(([, value]) => value !== undefined)) as T
}

export function resolveOptions(
  config: NuxtESLintConfigOptions,
): NuxtESLintConfigOptionsResolved {
  if ('__resolved' in config) {
    return config as NuxtESLintConfigOptionsResolved
  }

  const dirs = {
    ...config.dirs,
  } as NuxtESLintConfigOptionsResolved['dirs']

  dirs.root ||= ['.', './app'] // Support both Nuxt 3 and 4 conventions by default
  dirs.src ||= dirs.root
  dirs.pages ||= dirs.src.map(src => `${src}/pages`)
  dirs.layouts ||= dirs.src.map(src => `${src}/layouts`)
  dirs.components ||= dirs.src.map(src => `${src}/components`)
  dirs.composables ||= dirs.src.map(src => `${src}/composables`)
  dirs.plugins ||= dirs.src.map(src => `${src}/plugins`)
  dirs.modules ||= dirs.src.map(src => `${src}/modules`)
  dirs.middleware ||= dirs.src.map(src => `${src}/middleware`)
  dirs.servers ||= dirs.src.map(src => `${src}/servers`)
  dirs.componentsPrefixed ||= []

  const resolved: NuxtESLintConfigOptionsResolved = {
    features: {
      standalone: true,
      stylistic: false,
      typescript: isPackageExists('typescript'),
      tooling: false,
      formatters: false,
      nuxt: {},
      import: {},
      ...config.features,
    },
    dirs,
  }

  Object.defineProperty(resolved, '__resolved', { value: true, enumerable: false })

  return resolved
}
