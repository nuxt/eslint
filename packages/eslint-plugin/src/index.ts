import type { ESLint } from 'eslint'
import rules from './rules'

export default {
  meta: {
    name: 'nuxt-module-eslint-config/plugin',
  },
  rules,
} satisfies ESLint.Plugin
