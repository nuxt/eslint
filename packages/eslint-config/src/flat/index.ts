import type { ResolvableFlatConfig, FlatConfigComposer } from 'eslint-flat-config-utils'
import { composer } from 'eslint-flat-config-utils'
import gitignore from 'eslint-config-flat-gitignore'
import type { Linter } from 'eslint'
import type { NuxtESLintConfigOptions } from './types'
import disables from './configs/disables'
import nuxt from './configs/nuxt'
import ignores from './configs/ignores'
import javascript from './configs/javascript'
import { resolveOptions } from './utils'

export * from './types'

export { resolveOptions }

/**
 * Provide type definitions for constructing ESLint flat config items.
 *
 * This function takes flat config item, or an array of them as rest arguments.
 * It also automatically resolves the promise if the config item is a promise.
 */
export function defineFlatConfigs(
  ...configs: ResolvableFlatConfig[]
): FlatConfigComposer<Linter.FlatConfig> {
  return composer(...configs)
}

/**
 * Create an array of ESLint flat configs for Nuxt 3, based on the given options.
 * Accpets appending user configs as rest arguments from the second argument.
 *
 * For Nuxt apps, it's recommanded to use `@nuxt/eslint` module instead, which will generate the necessary configuration based on your project.
 * @see https://eslint.nuxt.com/packages/module
 */
export function createConfigForNuxt(
  options: NuxtESLintConfigOptions = {},
  ...userConfigs: ResolvableFlatConfig[]
): FlatConfigComposer<Linter.FlatConfig> {
  const c = composer()

  const resolved = resolveOptions(options)

  if (resolved.features.standalone !== false) {
    c.append(
      gitignore({ strict: false }),
      ignores(),
      javascript(),
      // Make these imports async, as they are optional and imports plugins
      import('./configs/typescript').then(m => m.default(resolved)),
      import('./configs/vue').then(m => m.default(resolved)),
      import('./configs/import').then(m => m.default(resolved)),
    )
  }

  c.append(
    nuxt(resolved),
  )

  if (resolved.features.tooling) {
    const toolingOptions = typeof resolved.features.tooling === 'boolean' ? {} : resolved.features.tooling
    c.append(
      toolingOptions.jsdoc !== false && import('./configs-tooling/jsdoc').then(m => m.default(resolved)),
      toolingOptions.unicorn !== false && import('./configs-tooling/unicorn').then(m => m.default()),
      toolingOptions.regexp !== false && import('./configs-tooling/regexp').then(m => m.default()),
    )
  }

  if (resolved.features.stylistic) {
    const stylisticOptions = typeof resolved.features.stylistic === 'boolean'
      ? {}
      : resolved.features.stylistic

    c.append(
      import('./configs/stylistic').then(m => m.default(stylisticOptions)),
    )
  }

  c.append(
    disables(resolved),
  )

  if (userConfigs.length > 0) {
    c.append(...userConfigs)
  }

  return c
}

export default createConfigForNuxt
