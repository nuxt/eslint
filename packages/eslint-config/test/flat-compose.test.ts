import { describe, expect, it } from 'vitest'
import type { FlatConfigItem } from 'eslint-flat-config-utils'
import { createConfigForNuxt } from '../src/flat'

const cwd = process.cwd()

function getFlatConfigDigest(configs: FlatConfigItem[]) {
  return configs.map((config) => {
    return JSON.parse(JSON.stringify({
      name: config.name,
      files: config.files,
      ignores: config.ignores,
    }).replaceAll(cwd, '<cwd>'))
  })
}

describe('flat config composition', () => {
  it('empty', async () => {
    const configs = await createConfigForNuxt()

    expect(getFlatConfigDigest(configs))
      .toMatchSnapshot()
  })

  it('non-standalone', async () => {
    const configs = await createConfigForNuxt({
      features: {
        standalone: false,
      },
    })

    expect(getFlatConfigDigest(configs))
      .toMatchSnapshot()
  })

  it('custom src dirs', async () => {
    const configs = await createConfigForNuxt({
      dirs: {
        src: ['src1', 'src2'],
      },
    })

    expect(getFlatConfigDigest(configs))
      .toMatchSnapshot()
  })
})
