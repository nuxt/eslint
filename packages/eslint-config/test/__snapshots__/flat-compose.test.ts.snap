// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`flat config composition > custom src dirs 1`] = `
[
  {
    "ignores": [
      "node_modules",
      "**/node_modules/**",
      "*.iml",
      "**/*.iml/**",
      ".idea",
      "**/.idea/**",
      "*.log",
      "**/*.log/**",
      ".nuxt",
      "**/.nuxt/**",
      ".output",
      "**/.output/**",
      "**/.yarn/cache",
      "**/.yarn/*state*",
      "dist",
      "**/dist/**",
      ".eslintcache",
      "**/.eslintcache/**",
    ],
  },
  {
    "ignores": [
      "**/dist",
      "**/node_modules",
      "**/.nuxt",
      "**/.output",
      "**/.vercel",
      "**/.netlify",
    ],
  },
  {
    "name": "nuxt/javascript",
  },
  {
    "name": "nuxt/typescript/setup",
  },
  {
    "files": [
      "**/*.ts",
      "**/*.tsx",
      "**/*.mts",
      "**/*.cts",
      "**/*.vue",
    ],
    "name": "nuxt/typescript/rules",
  },
  {
    "files": [
      "**/*.ts",
      "**/*.tsx",
      "**/*.mts",
      "**/*.cts",
      "**/*.vue",
    ],
    "name": "nuxt/typescript/disables",
  },
  {
    "name": "nuxt/vue/setup",
  },
  {
    "files": [
      "**/*.vue",
    ],
    "name": "nuxt/vue/rules",
  },
  {
    "files": [
      "src1/layouts/**/*.{js,ts,jsx,tsx,vue}",
      "src2/layouts/**/*.{js,ts,jsx,tsx,vue}",
      "src1/pages/**/*.{js,ts,jsx,tsx,vue}",
      "src2/pages/**/*.{js,ts,jsx,tsx,vue}",
    ],
    "name": "nuxt/vue/single-root",
  },
  {
    "name": "nuxt/rules",
  },
  {
    "files": [
      "src1/app.{js,ts,jsx,tsx,vue}",
      "src1/error.{js,ts,jsx,tsx,vue}",
      "src2/app.{js,ts,jsx,tsx,vue}",
      "src2/error.{js,ts,jsx,tsx,vue}",
      "src1/layouts/**/*.{js,ts,jsx,tsx,vue}",
      "src2/layouts/**/*.{js,ts,jsx,tsx,vue}",
      "src1/pages/**/*.{js,ts,jsx,tsx,vue}",
      "src2/pages/**/*.{js,ts,jsx,tsx,vue}",
      "src1/components/*/**/*.{js,ts,jsx,tsx,vue}",
      "src2/components/*/**/*.{js,ts,jsx,tsx,vue}",
    ],
    "name": "nuxt/disables/routes",
  },
]
`;

exports[`flat config composition > empty 1`] = `
[
  {
    "ignores": [
      "node_modules",
      "**/node_modules/**",
      "*.iml",
      "**/*.iml/**",
      ".idea",
      "**/.idea/**",
      "*.log",
      "**/*.log/**",
      ".nuxt",
      "**/.nuxt/**",
      ".output",
      "**/.output/**",
      "**/.yarn/cache",
      "**/.yarn/*state*",
      "dist",
      "**/dist/**",
      ".eslintcache",
      "**/.eslintcache/**",
    ],
  },
  {
    "ignores": [
      "**/dist",
      "**/node_modules",
      "**/.nuxt",
      "**/.output",
      "**/.vercel",
      "**/.netlify",
    ],
  },
  {
    "name": "nuxt/javascript",
  },
  {
    "name": "nuxt/typescript/setup",
  },
  {
    "files": [
      "**/*.ts",
      "**/*.tsx",
      "**/*.mts",
      "**/*.cts",
      "**/*.vue",
    ],
    "name": "nuxt/typescript/rules",
  },
  {
    "files": [
      "**/*.ts",
      "**/*.tsx",
      "**/*.mts",
      "**/*.cts",
      "**/*.vue",
    ],
    "name": "nuxt/typescript/disables",
  },
  {
    "name": "nuxt/vue/setup",
  },
  {
    "files": [
      "**/*.vue",
    ],
    "name": "nuxt/vue/rules",
  },
  {
    "files": [
      "layouts/**/*.{js,ts,jsx,tsx,vue}",
      "pages/**/*.{js,ts,jsx,tsx,vue}",
    ],
    "name": "nuxt/vue/single-root",
  },
  {
    "name": "nuxt/rules",
  },
  {
    "files": [
      "app.{js,ts,jsx,tsx,vue}",
      "error.{js,ts,jsx,tsx,vue}",
      "layouts/**/*.{js,ts,jsx,tsx,vue}",
      "pages/**/*.{js,ts,jsx,tsx,vue}",
      "components/*/**/*.{js,ts,jsx,tsx,vue}",
    ],
    "name": "nuxt/disables/routes",
  },
]
`;

exports[`flat config composition > non-standalone 1`] = `
[
  {
    "files": [
      "layouts/**/*.{js,ts,jsx,tsx,vue}",
      "pages/**/*.{js,ts,jsx,tsx,vue}",
    ],
    "name": "nuxt/vue/single-root",
  },
  {
    "name": "nuxt/rules",
  },
  {
    "files": [
      "app.{js,ts,jsx,tsx,vue}",
      "error.{js,ts,jsx,tsx,vue}",
      "layouts/**/*.{js,ts,jsx,tsx,vue}",
      "pages/**/*.{js,ts,jsx,tsx,vue}",
      "components/*/**/*.{js,ts,jsx,tsx,vue}",
    ],
    "name": "nuxt/disables/routes",
  },
]
`;
